File: ExpenseMaster/core/src/main/java/com/expensemaster/core/SpanAdapter.java
package com.expensemaster.core;

import com.expensemaster.application.ICallback;
import com.expensemaster.user.IUserSpan;
import io.opentelemetry.api.logs.LogRecordBuilder;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanKind;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Context;
import io.opentelemetry.context.Scope;
import io.opentelemetry.context.propagation.TextMapPropagator;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.amqp.core.MessageProperties;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Component
public class SpanAdapter implements IUserSpan, ISpanAdapter {
    private final Tracer tracer;
    private final TextMapPropagator textMapPropagator;
    private final LogRecordBuilder logRecordBuilder;
    private HttpServletRequest request;

    private Span span;

    public SpanAdapter(
            final Tracer tracer,
            final TextMapPropagator textMapPropagator,
            final LogRecordBuilder logRecordBuilder
            ) {
        this.logRecordBuilder = Objects.requireNonNull(logRecordBuilder);
        this.tracer = Objects.requireNonNull(tracer);
        this.textMapPropagator = Objects.requireNonNull(textMapPropagator);
    }

    private Context makeContext() {
        return textMapPropagator.extract(Context.current(), request, new HeaderGetter());
    }

    @Override
    public void startSpan(final String spanName, ICallback function) {
        final var contextResult = this.makeContext();
        this.span = tracer.spanBuilder(spanName)
                .setSpanKind(SpanKind.SERVER)
                .setParent(contextResult)
                .startSpan();
        final var currrentContext = Context.current().with(this.span);
        try (Scope scope = this.span.makeCurrent()) {
            try {
                logRecordBuilder.setContext(currrentContext).emit();
                function.apply();
            } catch (Exception e) {
                this.span.recordException(e);
                throw e;
            } finally {
                this.span.end(Instant.now());
            }
        }
    }

    @Override
    public void startSpanWithContext(String spanName, ICallback function) {
        final var clientSpan = tracer.spanBuilder(spanName)
                .setSpanKind(SpanKind.CLIENT)
                .setParent(Context.current().with(this.span))
                .startSpan();
        try (Scope scope = clientSpan.makeCurrent()) {
            try {
                function.apply();
            } catch (Exception e) {
                clientSpan.recordException(e);
                throw e;
            } finally {
                clientSpan.end(Instant.now());
            }
        }
    }

    @Override
    public MessageProperties contextPropagationQueue() {
        final var messageProperties = new MessageProperties();
        final var context = Context.current().with(this.span);

        this.textMapPropagator.inject(context, messageProperties.getHeaders(), (carrier, key, value) -> {
            carrier.put(key, value);
        });

        return messageProperties;
    }

    @Override
    public List<ClientHttpRequestInterceptor> contextPropagationApi() {
        List<ClientHttpRequestInterceptor> interceptors = new ArrayList<>();
        interceptors.add(new HeaderPropagationInterceptor());
        return interceptors;
    }

    @Override
    public void setHttpRequest(final HttpServletRequest request) {
        this.request = request;
    }
}

File: ExpenseCore/core/src/infra/SpanAdapter.ts
import { trace, Context, Tracer, Span, SpanKind, ROOT_CONTEXT, propagation } from '@opentelemetry/api';
import { Headers, ISpan } from "./ISpan";

export class SpanAdapter implements ISpan {
    [key: string]: any;
    private readonly SERVICE_NAME = "Expense_Core";
    private readonly SERVICE_VERSION = "0.0.1";
    private tracer: Tracer;
    private context?: Context;
    private spanWithContext?: Span;
    private spanWithoutContext?: Span;

    constructor(serviceName: string, version: string) {
        this.tracer = trace.getTracer(serviceName, version);
    }

    public setContext(headers: Headers): void {
        this.context = propagation.extract(ROOT_CONTEXT, headers)
    }

    public startSpanWithoutContext(spanName: string): void {
        this.spanWithoutContext = this.tracer.startSpan(spanName, { kind: SpanKind.SERVER, startTime: new Date() }, this.context as Context);
    }

    public endSpanWithoutContext(): void {
        this.spanWithoutContext!.end(new Date());
    }

    public startSpanWithContext(spanName: string): void {
        this.spanWithContext = this.tracer.startSpan(spanName, { kind: SpanKind.CLIENT, startTime: new Date() }, this.context!)
    }

    public endSpanWithContext(): void {
        this.spanWithContext!.end(new Date());
    }

    public getSpanServer(): Span {
        return this.spanWithoutContext!
    }

    public contextPropagationWith(): Record<string, string> {
        const context = trace.setSpanContext(ROOT_CONTEXT, this.spanWithContext!.spanContext())
        const headers: Record<string, string> = {}
        propagation.inject(context, headers)
        return headers
    }

}